frontend/ui/widget/textviewer.lua

```diff
@@ -0,0 +1,238 @@
+local UIManager = require("ui/uimanager")
+local VerticalGroup = require("ui/widget/verticalgroup")
+local WidgetContainer = require("ui/widget/container/widgetcontainer")
+local logger = require("logger")
+local _ = require("gettext")
+local Screen = Device.screen
+
+local TextViewer = InputContainer:new{
+    title = nil,
+    text = nil,
+    width = nil,
+    height = nil,
+
+    title_face = Font:getFace("x_smalltfont"),
+    text_face = Font:getFace("x_smallinfofont"),
+    title_padding = Screen:scaleBySize(5),
+    title_margin = Screen:scaleBySize(2),
+    text_padding = Screen:scaleBySize(10),
+    text_margin = Screen:scaleBySize(2),
+    button_padding = Screen:scaleBySize(14),
+}
+
+function TextViewer:init()
+    local orig_dimen = self.frame and self.frame.dimen or Geom:new{}
+    -- calculate window dimension
+    self.align = "center"
+    self.region = Geom:new{
+        x = 0, y = 0,
+        w = Screen:getWidth(),
+        h = Screen:getHeight(),
+    }
+    self.width = self.width or Screen:getWidth() - Screen:scaleBySize(30)
+    self.height = self.height or Screen:getHeight() - Screen:scaleBySize(30)
+
+    if Device:hasKeys() then
+        self.key_events = {
+            Close = { {"Back"}, doc = "close text viewer" }
+        }
+    end
+
+    if Device:isTouchDevice() then
+        self.ges_events = {
+            TapClose = {
+                GestureRange:new{
+                    ges = "tap",
+                    range = Geom:new{
+                        x = 0, y = 0,
+                        w = Screen:getWidth(),
+                        h = Screen:getHeight(),
+                    }
+                },
+            },
+            Swipe = {
+                GestureRange:new{
+                    ges = "swipe",
+                    range = Geom:new{
+                        x = 0, y = 0,
+                        w = Screen:getWidth(),
+                        h = Screen:getHeight(),
+                    }
+                },
+            },
+        }
+    end
+
+    local title_text = TextWidget:new{
+        text = self.title,
+        face = self.title_face,
+        bold = true,
+        width = self.width - 2*self.title_padding - 2*self.title_margin,
+    }
+    local titlew = FrameContainer:new{
+        padding = self.title_padding,
+        margin = self.title_margin,
+        bordersize = 0,
+        CenterContainer:new{
+            dimen = Geom:new{
+                w = self.width,
+                h = title_text:getSize().h,
+            },
+            title_text
+        }
+    }
+    titlew = OverlapGroup:new{
+        dimen = {
+            w = self.width,
+            h = titlew:getSize().h
+        },
+        titlew,
+        CloseButton:new{ window = self, },
+    }
+
+    local separator = LineWidget:new{
+        dimen = Geom:new{
+            w = self.width,
+            h = Screen:scaleBySize(2),
+        }
+    }
+
+    local buttons = {
+        {
```

Comment:
Again, I followed what is done in most other existing code I took inspiration from, sticking to tradition :)